### PR TITLE
Add a double-precision editor build to CI

### PR DESCRIPTION
**Pull request by aaronfranke**
This PR fixes enough of Godot's code so that it compiles with `float=64` again and adds a double-precision editor build to the CI. Some of the code changes are for the OpenGL support, some are in the text server, and some are in GLTF. Once this is in CI, this will check `float=64` for every commit so that it will keep compiling.

### CODE REVIEW

**Review by fire — APPROVED** [review]
This is required to keep float=64 from breaking.

**Comment by fire** [comment]
A common request is to combine it with one of the existing cicd builds, but will wait for that request.

**Comment by fire** [comment]
> `<Akien>` Adds another build to CI, which will make CI slower. I haven't assessed in detail yet. 

I think we need to keep the number of builds the same.

**Comment by Calinou** [comment]
For nonstandard builds, I would probably look at moving them to a separate CI workflow that run according to a weekly schedule. This way, we don't have to worry about stealing CI time from other commits and pull requests.

**Comment by fire** [comment]
The point is to prevent merges that fail this test and have the contributors attend to the breakage. Having them be non fatal would defeat this goal.

**Review by bruvzg — APPROVED** [review]
`TextServer` still have a few more `real_t` variables used, it won't break a build, but should be eventually changed to avoid unnecessary conversions.

**Comment by aaronfranke** [comment]
@akien-mga Done, it's now modifying that build.

File: scene/resources/font.cpp [hunk 1]
```diff
@@ -1551,7 +1551,7 @@ void Font::draw_string(RID p_canvas_item, const Point2 &p_pos, const String &p_t
 	buffer->draw(p_canvas_item, ofs, p_modulate);
 }
 
-void Font::draw_multiline_string(RID p_canvas_item, const Point2 &p_pos, const String &p_text, HAlign p_align, float p_width, int p_max_lines, int p_size, const Color &p_modulate, int p_outline_size, const Color &p_outline_modulate, uint16_t p_flags) const {
+void Font::draw_multiline_string(RID p_canvas_item, const Point2 &p_pos, const String &p_text, HAlign p_align, real_t p_width, int p_max_lines, int p_size, const Color &p_modulate, int p_outline_size, const Color &p_outline_modulate, uint16_t p_flags) const {
```
Comment (akien-mga) [inline scene/resources/font.cpp:1554]: I see here you're changing `float` to `real_t` for the width parameter, but in TextServer it's changed the other way around. How do you decide? (In both cases it's small numbers a very high precision pixel values don't really make sense so `float` would be good enough precision wise.)

File: scene/resources/importer_mesh.cpp [hunk 2]
```diff
@@ -517,10 +517,10 @@ void ImporterMesh::generate_lods(float p_normal_merge_angle, float p_normal_spli
 					Vector3 normal = n0 * w + n1 * u + n2 * v;
 
 					Vector2 orig_uv = ray_uvs[j];
-					float orig_bary[3] = { 1.0f - orig_uv.x - orig_uv.y, orig_uv.x, orig_uv.y };
+					real_t orig_bary[3] = { 1.0f - orig_uv.x - orig_uv.y, orig_uv.x, orig_uv.y };
```
Comment (akien-mga) [inline scene/resources/importer_mesh.cpp:520]: Should the `1.0f` be changed?

File: scene/3d/cpu_particles_3d.cpp [hunk 3]
```diff
@@ -775,7 +775,7 @@ void CPUParticles3D::_particles_process(double p_delta) {
 				binormal.normalize();
 				Vector3 normal = binormal.cross(direction_nrm);
 				spread_direction = binormal * spread_direction.x + normal * spread_direction.y + direction_nrm * spread_direction.z;
-				p.velocity = spread_direction * Math::lerp(parameters_min[PARAM_INITIAL_LINEAR_VELOCITY], parameters_max[PARAM_INITIAL_LINEAR_VELOCITY], float(Math::randf()));
+				p.velocity = spread_direction * Math::lerp(parameters_min[PARAM_INITIAL_LINEAR_VELOCITY], parameters_max[PARAM_INITIAL_LINEAR_VELOCITY], real_t(Math::randf()));
```
Comment (akien-mga) [inline scene/3d/cpu_particles_3d.cpp:778]: Could be `(real_t)` cast style to be consistent.

File: scene/resources/importer_mesh.cpp (hunk 2 above)
Comment (aaronfranke) [reply scene/resources/importer_mesh.cpp:520]: It doesn't need to, it has the same behavior as `(real_t)1.0` or similar. `float - double` will implicitly make the float become double, and `1.0f` is represented exactly.

File: scene/resources/font.cpp (hunk 1 above)
Comment (aaronfranke) [reply scene/resources/font.cpp:1554]: I was going for the minimum to make it compile, but it's better to just make this float.

**Comment by akien-mga** [comment]
Thanks!

**Comment by Calinou** [comment]
Floats in GDScript are always double-precision, but packed arrays can store either 32-bit or 64-bit floats in Godot 4.0. In Godot 3.x, PoolRealArray always stores 32-bit floats.

**Comment by HeadClot** [comment]
> _Scalar_ floats in GDScript are always double-precision, but packed arrays can store either 32-bit or 64-bit floats in Godot 4.0 (PackedFloat32Array vs PackedFloat64Array). In Godot 3.x, PoolRealArray always stores 32-bit floats. The same thing applies to packed/pooled integer arrays.

Ok good to know. Thank you for the very insightful reply :)

**Pull request merged**